Automattic/jetpack
# Minimum Sync Config: Update required modules and options

This PR updates the minimum required data settings for Sync while fixing a few existing logic errors.

#### Changes proposed in this Pull Request:
Fixes:
* `Sync\Data_Settings::add_filters_custom_settings_and_hooks`: Updated the check for enabled modules to also take into account the ones defined under `MUST_SYNC_DATA_SETTINGS`. This fixes an inconsistency, where if the corresponding module was not provided by the consumer plugin under `jetpack_sync_modules` it would be activated with the default settings otherwise with the minimum settings. Covered in the following unit test: `Data_Test_Data_Settings::data_test_1_2`
* `Sync\Data_Settings::add_required_settings`: No need to add a custom filter setting if the corresponding setting is already set and matches the default one.

`Data_Settings::MUST_SYNC_DATA_SETTINGS`: 
* Add `Options` and `Constants` to the list of required modules
* Update required `jetpack_sync_callable_whitelist`
* Add required `jetpack_sync_constants_whitelist`
* Add required `jetpack_sync_options_whitelist`

README:
* Detailed instructions on getting started with Sync and configuring what data to sync.


#### Other information:


#### Jetpack product discussion


p9dueE-5lh-p2

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Get a test site with any Jetpack plugin that uses Sync (Jetpack, Backup).
* Connect site, observe Sync activity - see the full sync and all added data being synced correctly.
* Change the settings to minimal using the Config object:

```php

    use Automattic\Jetpack\Sync\Data_Settings;

    // ....
    // Change the existing `ensure` call to this: 
    $config->ensure( 'sync', Data_Settings::MUST_SYNC_DATA_SETTINGS );
```

* Observe that data doesn't get synced and you can only see options, callables and constants getting synced.
* Try again with multiple plugins installed, see that the settings get combined, not overwritten.
